### PR TITLE
Add CUDA Kernels of RandomNormal[Like], RandomUniform[Like]

### PR DESCRIPTION
**Pull request by iK1D**
Add CUDA kernels for RandomNormal, RandomNormalLike, RandomUniform, RandomUniformLike.

**Motivation and Context**
- MoE model requires RandomUniformLike.

### CODE REVIEW

File: onnxruntime/core/providers/cuda/generator/random.cc [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+Status RandomNormal::ComputeInternal(OpKernelContext* p_ctx) const { return Compute(p_ctx, shape_, dtype_); }
+
+Status RandomNormalLike::ComputeInternal(OpKernelContext* p_ctx) const {
+  const Tensor* p_X = p_ctx->Input<Tensor>(0);
+  if (!p_X) return Status(common::ONNXRUNTIME, common::FAIL, "X Input is not available.");
+  const TensorShape& shape = p_X->Shape();
+  auto dtype = dtype_ != TensorProto_DataType_UNDEFINED ? dtype_ : InferDataType(*p_X);
+  if (dtype == TensorProto_DataType_UNDEFINED) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Could not infer data type from input tensor with data type ",
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/generator/random.cc:62]: >Could not infer data type from input tensor with data type [](http://example.com/codeflow?start=47&length=58)

This is a very confusing error message. It makes it look like the data type is absent. The runtime enforces supported data types. The InferDataType*() function really filters for specific data types. You can simply use IsDataType<T>() methods if you need to, that would be much more readable.

File: onnxruntime/core/providers/cuda/generator/random.cc [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                           p_X->DataType());
+  }
+  return Compute(p_ctx, shape, dtype);
+}
+
+Status RandomUniformBase::Compute(OpKernelContext* p_ctx, const TensorShape& shape, int dtype) const {
+  Tensor& Y = *p_ctx->Output(0, shape);
+  const int64_t N = shape.Size();
+  PhiloxGenerator& generator = generator_ ? *generator_ : PhiloxGenerator::Default();
+  utils::MLTypeCallDispatcher<float, MLFloat16, double> t_disp(dtype);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cuda/generator/random.cc:72]: >float, MLFloat16, double [](http://example.com/codeflow?start=30&length=24)

Declare supported types as a type list in one place and use it for kernel declarations and MLTypeDispatcher.

File: onnxruntime/test/providers/cpu/generator/random_test.cc [hunk 3]
```diff
@@ -33,7 +33,7 @@ TEST(Random, RandomNormal2DDouble) {
                 [&generator, &distribution](double& value) { value = distribution(generator); });
 
   test.AddOutput<double>("Y", dims, expected_output);
-  test.Run();
+  test.Run(OpTester::ExpectResult::kExpectSuccess, "", {kCudaExecutionProvider});
```
Comment (SherlockNoMad) [inline onnxruntime/test/providers/cpu/generator/random_test.cc:36]: I am curious, why these test are not passing for CUDA EP?

Comment (iK1D) [reply onnxruntime/test/providers/cpu/generator/random_test.cc:36]: I used std lib to generate the expected output, which is used for CPU kernel only.

  std::default_random_engine generator{gsl::narrow_cast<uint32_t>(seed)};
  std::normal_distribution<double> distribution{mean, scale};

File: onnxruntime/core/providers/cuda/generator/random.h [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+RANDOM_COMPUTE_IMPL(RandomNormal)
+RANDOM_COMPUTE_IMPL(RandomUniform)
+
+#undef RANDOM_COMPUTE_IMPL
+
+class RandomBase : public CudaKernel {
+ protected:
+  RandomBase(const OpKernelInfo& info) : CudaKernel(info) {
+    float seed = 0.f;
+    if (info.GetAttr<float>("seed", &seed).IsOK()) {
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/generator/random.h:33]: we need to use the global seed, if the seed in attribute is not set. 
See Random CPU kernel for reference.

File: onnxruntime/core/providers/cuda/generator/random.h [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    if (info.GetAttr<int64_t>("dtype", &dtype).IsOK()) {
+      dtype_ = static_cast<ONNX_NAMESPACE::TensorProto::DataType>(dtype);
+      ORT_ENFORCE(
+          ONNX_NAMESPACE::TensorProto::DataType_IsValid(dtype_) && dtype_ != ONNX_NAMESPACE::TensorProto::UNDEFINED,
+          "Invalid dtype of ", dtype_);
+    }
+  }
+
+ protected:
+  mutable std::unique_ptr<PhiloxGenerator> generator_;
```
[the remark this answers is not in the document]
Comment (SherlockNoMad) [reply onnxruntime/core/providers/cuda/generator/random.h:47]: this is a random generator, so it needs to be stateful.

File: onnxruntime/core/providers/cuda/generator/random.h [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+ protected:
+  RandomBase(const OpKernelInfo& info) : CudaKernel(info) {
+    float seed = 0.f;
+    if (info.GetAttr<float>("seed", &seed).IsOK()) {
+      generator_ = std::make_unique<PhiloxGenerator>(static_cast<uint64_t>(seed));
+    }
+
+    int64_t dtype;
+    if (info.GetAttr<int64_t>("dtype", &dtype).IsOK()) {
+      dtype_ = static_cast<ONNX_NAMESPACE::TensorProto::DataType>(dtype);
```
Comment (SherlockNoMad) [inline onnxruntime/core/providers/cuda/generator/random.h:39]: According to onnx spec, default is TensorProto::FLOAT.

Comment (SherlockNoMad) [reply onnxruntime/core/providers/cuda/generator/random.h:39]: For random, the default is Float. For randomLike, there is no default.

File: onnxruntime/core/providers/cuda/generator/random.h (hunk 4 above)
Comment (iK1D) [reply onnxruntime/core/providers/cuda/generator/random.h:33]: When seed attr is not set, we will use default generator, which use the global seed, and it's static, so different nodes will generate different data. Dropout is also doing this way.

PhiloxGenerator& generator = generator_ ? *generator_ : PhiloxGenerator::Default();

File: onnxruntime/core/providers/cuda/generator/random.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnxruntime/core/providers/cuda/generator/random.h:47]: Well, yes, but do we run the model from multiple threads? Plus, mutable keyword here applies to the smart pointer, not the actual generator object.

[the remark this answers is not in the document]
Comment (pranavsharma) [reply onnxruntime/core/providers/cuda/generator/random.h:47]: generator_ should be protected by a mutex to account for multithreading scenarios. Also, can the heap allocation be avoided?

File: onnxruntime/test/providers/cpu/generator/random_test.cc (hunk 3 above)
Comment (SherlockNoMad) [reply onnxruntime/test/providers/cpu/generator/random_test.cc]: got it. let's add this reason as a comment.

File: onnxruntime/core/providers/cuda/generator/random.h (hunk 4 above)
Comment (SherlockNoMad) [reply onnxruntime/core/providers/cuda/generator/random.h:33]: got it.

File: onnxruntime/core/providers/cuda/generator/random.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (iK1D) [reply onnxruntime/core/providers/cuda/generator/random.h]: I think yes the mutable is not necessary here. We never change the smart ptr here, and when access it we use `operator*() const` which returns non-const reference. And we already have the mutex to protect inside the PhiloxGenerator.

**Review by SherlockNoMad — APPROVED** [review]

**Pull request merged**